nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -1583,6 +1626,274 @@ impl DataStore {
+        // table look like this prior to the transaction:
+        //
+        //            id | volume_id
+        //  -------------| ---------
+        //    TARGET_REG | TARGET_VOL
+        //       NEW_REG | NEW_VOL
+        //
+        // TARGET_VOL has a volume construction request where one of the targets
+        // list will contain TARGET_REG's address:
+        //
+        //   {
+        //     "type": "volume",
+        //     "block_size": 512,
+        //     "id": "TARGET_VOL",
+        //     "read_only_parent": {
+        //       ...
+        //     },
+        //     "sub_volumes": [
+        //       {
+        //         ...
+        //         "opts": {
+        //           ...
+        //           "target": [
+        //             "[fd00:1122:3344:103::3]:19004",
+        //             "[fd00:1122:3344:79::12]:27015",
+        //             "[fd00:1122:3344:145::10]:40001"  <-----
+        //           ]
+        //         }
+        //       }
+        //     ]
+        //   }
+        //
+        // Note it is not required for the replacement volume to exist as a
+        // database record for this transaction.
+        //
+        // The first part of the transaction will swap the volume IDs of the
+        // target and replacement region records:
+        //
+        //           id | volume_id
+        //  ------------| ---------
+        //   TARGET_REG | NEW_VOL
+        //      NEW_REG | TARGET_VOL
+        //
+        // The second part of the transaction will update the volume
+        // construction request of TARGET_VOL by finding and replacing
+        // TARGET_REG's address (in the appropriate targets array) with
+        // NEW_REG's address:
+        //
+        //   {
+        //           ...
+        //           "target": [
+        //             "[fd00:1122:3344:103::3]:19004",
+        //             "[fd00:1122:3344:79::12]:27015",
+        //             "[fd00:1122:3344:322::4]:3956"  <-----
+        //           ]
+        //           ...
+        //   }
+        //
+        // After the transaction, the caller should ensure that TARGET_REG is
+        // referenced (via its socket address) in NEW_VOL. For an example, this
+        // is done as part of the region replacement start saga.
+
+        #[derive(Debug, thiserror::Error)]
+        enum VolumeReplaceRegionError {
+            #[error("Error from Volume region replacement: {0}")]
+            Public(Error),
+
+            #[error("Serde error during Volume region replacement: {0}")]
+            SerdeError(#[from] serde_json::Error),
+
+            #[error("Target Volume deleted")]
+            TargetVolumeDeleted,
+
+            #[error("Region replacement error: {0}")]
+            RegionReplacementError(#[from] anyhow::Error),
+        }
+        let err = OptionalError::new();
+
+        let conn = self.pool_connection_unauthorized().await?;
+        self.transaction_retry_wrapper("volume_replace_region")
```

Comment:
Can this run forever? Is there a way to mitigate that?

Reply:
I don't think so. In terms of other operations that can modify a volume construction request, it can race with the removal of a read-only parent, but if that's the case it will retry.